crates/matrix-sdk/src/widget_api/driver/mod.rs

```diff
@@ -1,100 +1,108 @@
+
     async fn get_openid(&self, req: openid::Request) -> OpenIDState {
-        // TODO: make the client ask the user first.
-        // We wont care about this for Element call -> Element X
-        // if !self.has_open_id_user_permission() {
-        //     let (rx,tx) = tokio::oneshot::channel();
-        //     return OpenIDState::Pending(tx);
-        //     let permissions = widget.openid_permissions().await?;
-        //     if (permissions.should_cache){
-        //        self.auto_allow_openid_token(permission.is_allowed);
-        //     }
-        //     self.get_openid(req, Some(tx)); // get open id can be called with or without tx and either reutrns as return or sends return val over tx
-        // }
-
-        let user_id = self.matrix_room.client.user_id();
-        if user_id == None {
-            return OpenIDState::Resolved(Err(WidgetError(
-                "Failed to get an open id token from the homeserver. Because the userId is not available".to_owned()
-            )));
-        }
-        let user_id = user_id.unwrap();
-
-        let request =
-            ruma::api::client::account::request_openid_token::v3::Request::new(user_id.to_owned());
-        let res = self.matrix_room.client.send(request, None).await;
-
-        let state = match res {
-            Err(err) => Err(WidgetError(
-                format!(
-                    "Failed to get an open id token from the homeserver. Because of Http Error: {}",
-                    err.to_string()
-                )
-                .to_owned(),
-            )),
-            Ok(res) => Ok(openid::Response {
-                id: req.id,
-                token: res.access_token,
-                expires_in_seconds: res.expires_in.as_secs() as usize,
-                server: res.matrix_server_name.to_string(),
-                kind: res.token_type.to_string(),
-            }),
-        };
+        let user_id = self.room.own_user_id();
+
+        let request = OpenIDRequest::new(user_id.to_owned());
+        let state = self.room.client.send(request, None).await.ok().map(|res| openid::Response {
+            id: req.id,
+            token: res.access_token,
+            expires_in_seconds: res.expires_in.as_secs() as usize,
+            server: res.matrix_server_name.to_string(),
+            kind: res.token_type.to_string(),
+        });
+
         OpenIDState::Resolved(state)
     }
 }
 
+impl<W> Driver<W> {
+    fn setup_event_listener(&mut self, filter: Filters) -> mpsc::UnboundedReceiver<MatrixEvent> {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let callback = move |ev: Raw<AnySyncTimelineEvent>| {
+            let (filter, tx) = (filter.clone(), tx.clone());
+            if let Ok(msg) = ev.deserialize_as::<MatrixEvent>() {
+                filter.allow(&msg).then(|| tx.send(msg));
+            }
+            async {}
+        };
+
+        self.event_handler_handle = Some(self.room.add_event_handler(callback));
+        return rx;
+    }
+}
+
```

Comment:
copied over from: https://github.com/toger5/matrix-rust-sdk/pull/20/files#diff-febc64d55a9708f4d2cd63ac427fdd5ef55ae5b3dd9c45c4c59d2b9de1537ba6
```suggestion
impl<W: Widget> Drop for Driver<W> {
    fn drop(&mut self) {
      self.event_handler_handle.map(|h| self.room.client().remove_event_handler(h.clone())
    }
}
```

Reply:
Good catch! I did not know that it requires doing this action manually, but luckily we have an [`EventHandlerDropGuard`]([EventHandlerDropGuard](https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk/event_handler/struct.EventHandlerDropGuard.html#)) for that! This is something we want to use given that we also must do the same action each time listener is set up.

Reply:
oh that is even bettern then.

Reply:
Done.